File: problems/11_container-with-most-water/main.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int left = 0;
        int right = height.size() - 1;
        int max_area = 0;
        while (left < right)
        {
            int h = min(height[left], height[right]);
            int w = right - left;
            max_area = max(max_area, h * w);
            if (height[left] < height[right])
            {
                left++;
            }
            else
            {
                right--;
            }
        }
        return max_area;
    }

    int maxArea2(vector<int> &height)
    {
        int maxArea = INT_MIN;
        int i = 0, j = height.size() - 1;
        while (i < j)
        {
            int currArea = (j - i) * min(height[i], height[j]);
            maxArea = max(maxArea, currArea);
            if (height[i] <= height[j])
            {
                i++;
            }
            else
            {
                j--;
            }
        }
        return maxArea;
    }
};

int main()
{
    Solution solution;
    vector<vector<int>> samples = {
        {1, 8, 6, 2, 5, 4, 8, 3, 7},
        {1, 1}};
    vector<int> expected = {49, 1};

    for (size_t i = 0; i < samples.size(); ++i)
    {
        vector<int> height = samples[i];
        int result = solution.maxArea(height);
        cout << "Sample " << (i + 1) << ": result=" << result
             << ", expected=" << expected[i] << endl;
    }

    return 0;
}

File: problems/11_container-with-most-water/main.rs
struct Solution;

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        if height.len() < 2 {
            return 0;
        }
        let mut left: usize = 0;
        let mut right: usize = height.len() - 1;
        let mut max_area: i32 = 0;

        while left < right {
            let h = height[left].min(height[right]);
            let w = (right - left) as i32;
            max_area = max_area.max(h * w);

            if height[left] < height[right] {
                left += 1;
            } else {
                right -= 1;
            }
        }

        max_area
    }

    pub fn max_area2(height: Vec<i32>) -> i32 {
        if height.len() < 2 {
            return 0;
        }
        let mut max_area = i32::MIN;
        let mut i: usize = 0;
        let mut j: usize = height.len() - 1;

        while i < j {
            let curr_area = ((j - i) as i32) * height[i].min(height[j]);
            max_area = max_area.max(curr_area);
            if height[i] <= height[j] {
                i += 1;
            } else {
                j -= 1;
            }
        }

        max_area
    }
}

fn main() {
    let samples = vec![
        (vec![1, 8, 6, 2, 5, 4, 8, 3, 7], 49),
        (vec![1, 1], 1),
    ];

    for (idx, (input, expected)) in samples.into_iter().enumerate() {
        let r1 = Solution::max_area(input.clone());
        let r2 = Solution::max_area2(input);
        println!(
            "Sample {}: max_area={}, max_area2={}, expected={}",
            idx + 1,
            r1,
            r2,
            expected
        );
    }
}
